spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/writer/GroupingBatchBuilder.scala

```diff
@@ -0,0 +1,122 @@
+                              batchKeyGenerator: BoundStatement => Any,
+                              batchSize: BatchSize,
+                              maxBatches: Int,
+                              data: Iterator[T]) extends AbstractIterator[Statement] with Iterator[Statement] {
+  require(maxBatches > 0)
+
+  private[this] val batchMap = new PriorityHashMap[Any, Batch](maxBatches)
+  private[this] val emptyBatches = new util.Stack[Batch]()
+  emptyBatches.ensureCapacity(maxBatches + 1)
+
+  private[this] var lastStatement: BoundStatement = null
+
+  // last key computation could be expensive, especially it would be a token range; therefore it is better
+  // to store the value in a temporary variable for subsequent usages
+  private[this] var lastKey: Any = null
+
+  /** The contract of this method is to return `Some(Batch)` if the returned batch needs to be consumed
+    * first, before the given statement can be processed, that is, the statement is not processed in this
+    * case. The method returns `None` if the statement is successfully added to the batch in the queue. */
+  private def processStatement(batchKey: Any, boundStatement: BoundStatement): Option[Batch] = {
+    batchMap.get(batchKey) match {
+      case Some(batch) =>
+        updateBatchInMap(batchKey, batch, boundStatement)
+      case None =>
+        addBatchToMap(batchKey, boundStatement)
+    }
+  }
+
+  private def updateBatchInMap(batchKey: Any, batch: Batch, newStatement: BoundStatement): Option[Batch] = {
+    if (batch.add(newStatement, force = false)) {
+      batchMap.put(batchKey, batch)
+      None
+    } else {
+      batchMap.remove(batchKey)
+      Some(batch)
+    }
+  }
+
+  private def addBatchToMap(batchKey: Any, newStatement: BoundStatement): Option[Batch] = {
+    if (batchMap.size == maxBatches) {
+      Some(batchMap.dequeue())
+    } else {
+      val batch = newBatch()
+      batch.add(newStatement, force = true)
+      batchMap.put(batchKey, batch)
+      None
+    }
+  }
+
+  /** Tries to get unused batch from the stack; if the stack is empty, it creates a new `Batch` object */
+  private def newBatch(): Batch = {
+    if (emptyBatches.isEmpty)
+      Batch(batchSize)
+    else
+      emptyBatches.pop()
+  }
+
+  /** Creates a statement from `Batch` object, clears the batch and puts it back on the stack */
+  private def createStmtAndReleaseBatch(batch: Batch): Statement = {
+    val stmt = batchStatementBuilder.maybeCreateBatch(batch.statements)
+    batch.clear()
+    emptyBatches.push(batch)
+    stmt
+  }
+
+  @tailrec
+  final override def computeNext(): Statement = {
+    if (lastStatement == null && data.hasNext) {
+      lastStatement = batchStatementBuilder.bind(data.next())
+      lastKey = batchKeyGenerator(lastStatement)
+    }
+
+    if (lastStatement != null) {
+      processStatement(lastKey, lastStatement) match {
+        case Some(batch) =>
+          createStmtAndReleaseBatch(batch)
```

Comment:
This code is quite non-obvious and I remember the previous one was much easier to understand. E.g. it is quite confusing for me that `processStatement` may fail and not add the statement, and the method `createStmtAndReleaseBatch` doesn't do anything with the not-added statement, yet the statement processing is (silently) postponed to the next iteration by *not setting* `lastStatement` to null. So it requires me to actually analyze not just what was written, but also what was *not written*.

I wonder, why can't we do it similar to as it was before:
1. grab the row from `data` and create the statement
2. create a batch key
3. find a batch or create a new one
4. add a statement to the batch
5. update priority map (call put)
6. the batch is full (exceeded) after addition? remove from the `PriorityMap` and return it
7. the number of batches is exceeded? take the first one from the `PriorityMap` and return it
8. nothing to return? continue `computeNext` tail call